src/software/networking/proto_multicast_listener.tpp

```diff
@@ -8,26 +8,83 @@ template <class ReceiveProtoT>
+    }
+
+    // Join the multicast group.
+    socket_.set_option(boost::asio::ip::multicast::join_group(
+        boost::asio::ip::address::from_string(ip_address)));
+
+    // Start listening for data asynchronously
+    // See here for a great explanation about asynchronous operations:
+    // https://stackoverflow.com/questions/34680985/what-is-the-difference-between-asynchronous-programming-and-multithreading
+    socket_.async_receive_from(boost::asio::buffer(raw_received_data_, max_buffer_length),
+                               sender_endpoint_,
+                               boost::bind(&ProtoMulticastListener::handleDataReception,
+                                           this, boost::asio::placeholders::error,
+                                           boost::asio::placeholders::bytes_transferred));
 }
 
 template <class ReceiveProtoT>
 void ProtoMulticastListener<ReceiveProtoT>::handleDataReception(
-    std::vector<uint8_t>& data)
+    const boost::system::error_code& error, size_t num_bytes_received)
 {
-    ReceiveProtoT packet_data;
-    const bool parsing_succeeded =
-        packet_data.ParseFromArray(data.data(), static_cast<int>(data.size()));
-    if (!parsing_succeeded)
+    if (!error)
     {
-        LOG(WARNING) << "Failed to parse received network packet into a "
-                     << TYPENAME(ReceiveProtoT);
-        return;
+        auto packet_data = ReceiveProtoT();
+        packet_data.ParseFromArray(raw_received_data_.data(),
+                                   static_cast<int>(num_bytes_received));
+        receive_callback(packet_data);
+
+        // Once we've handled the data, start listening again
+        socket_.async_receive_from(
+            boost::asio::buffer(raw_received_data_, max_buffer_length), sender_endpoint_,
+            boost::bind(&ProtoMulticastListener::handleDataReception, this,
+                        boost::asio::placeholders::error,
+                        boost::asio::placeholders::bytes_transferred));
     }
-    receive_callback(packet_data);
+    else
+    {
+        // Start listening again to receive the next data
+        socket_.async_receive_from(
+            boost::asio::buffer(raw_received_data_, max_buffer_length), sender_endpoint_,
+            boost::bind(&ProtoMulticastListener::handleDataReception, this,
+                        boost::asio::placeholders::error,
+                        boost::asio::placeholders::bytes_transferred));
+
+        LOG(WARNING)
+            << "An unknown network error occurred when attempting to receive ReceiveProtoT Data. The boost system error code is "
+            << error << std::endl;
+    }
+
+    if (num_bytes_received <= max_buffer_length)
+    {
+        LOG(WARNING)
+            << "num_bytes_received <= max_buffer_length, "
+            << "which means that the receive buffer is full and data loss has potentially occurred. "
+            << "Consider increasing max_buffer_length";
+    }
+}
```

Comment:
The logic here is inverted, we would want to log a warning when the num_bytes_received is greater than the max_buffer_length

Reply:
Done.